### DISCUSSION TITLE
Version 0.8.0-beta.1 test - ENDED

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#115] [Announcements]
**This is not a stable release, it's only meant for people to test new features. Do not use any important vault for testing, because you are not able to convert your data.json settings file back to an earlier version, and your shell commands might break and do strange things!**

If you want to start testing this, please read instructions on [Participating in beta testing](https://github.com/Taitava/obsidian-shellcommands/discussions/98) thread. Please give feedback regarding the beta installation process/installation issues in that topic! 🙂 

### Added
- [Autocompletion menu for variables in shell command fields (#33)](https://github.com/Taitava/obsidian-shellcommands/issues/33): When typing `{{` in a shell command field, a dropdown menu will open up, suggesting variables that match the input after `{{`. Variable help texts are also displayed. Extra: custom suggestions can be defined in an optional `autocomplete.yaml` file in the plugin's folder. It can contain e.g. all commands supported by a certain shell. TODO: Add documentation link here. The autocomplete feature is powered by [kraaden/autocomplete](https://github.com/kraaden/autocomplete).
- [Show SC version number in settings, and a link to the changelog (#112)](https://github.com/Taitava/obsidian-shellcommands/issues/112).

### Changed
- [Settings: Variables are now located on their own tab (#110)](https://github.com/Taitava/obsidian-shellcommands/issues/110).
- [Support ignoring error code 0 (#107)](https://github.com/Taitava/obsidian-shellcommands/issues/107)
- [Internal: Save new main settings fields to file already during loading (#111)](https://github.com/Taitava/obsidian-shellcommands/issues/111).

Please report any possible problems in this discussion (if they relate to the above-mentioned features). No need to create new discussions or issues for problems that are related to an ongoing beta test.

You can also drop a line if you think that everything is working fine in the beta! 🌻 

<hr /><em>This discussion was created from the release <a href='https://github.com/Taitava/obsidian-shellcommands/releases/tag/0.8.0-beta.1'>0.8.0-beta.1</a>.</em>

**Comment by Taitava** [comment]
# ~~Bug: The plugin does not load if a settings file does not exist~~ (edit: fixed in 0.8.0-beta.2)
If you are testing this on a vault that does not yet have an earlier version of SC installed, then the plugin won't have a `data.json` file available. You need to copy `.obsidian/plugins/obsidian-shellcommands/data.json` settings file from some other vault to the vault that you are using for the beta test. Then restart Obsidian and the beta version of SC should work ok.

I will fix this bug later and make `0.8.0-beta.2` test version.

Here's an error message from console: (console can be opened by pressing Ctrl/Cmd + Shift + I)
> Uncaught (in promise) TypeError: Cannot read property 'settings_version' of null
    at EnsureMainFieldsExist (eval at <anonymous> (app.js:1), <anonymous>:987:35)
    at RunMigrations (eval at <anonymous> (app.js:1), <anonymous>:886:9)
    at ShellCommandsPlugin.onload (eval at <anonymous> (app.js:1), <anonymous>:3013:15)

**Reply by Taitava** [reply]
# 0.8.0-beta.2 is available

The second beta fixes:
 - The loading bug if a settings file does not exist
 - Variables `{{file_path}}` and `{{folder_path}}` were not shown in the autocomplete menu.

**Comment by Taitava** [comment]
@FelipeRearden, when you have time, can you try out the autocomplete menu and see if there's anything you'd like to be changed. Thanks! No hurries, though! 🙂

**Comment by FelipeRearden** [comment]
> @FelipeRearden, when you have time, can you try out the autocomplete menu and see if there's anything you'd like to be changed. Thanks! No hurries, though! 🙂 


Sure @Taitava !!!!! 

Let's go !!!!!!!!!!

> Autocompletion menu for variables in shell command fields (#33): When typing {{ in a shell command field, a dropdown menu will open up, suggesting variables that match the input after {{. Variable help texts are also displayed. 

Perfect @Taitava !!!!!! AMAZING !!!!!!

I love the attention that you have with {{tags:separator}} to insert the caret position inside the }} instead of outside for variables with separator 😍 

> Extra: custom suggestions can be defined in an optional autocomplete.yaml file in the plugin's folder.

1- Show Autocomplete menu -> link `info` button

The link from this button https://publish.obsidian.md/shellcommands/Variables/Autocomplete/Autocomplete is 404 File Not Found 

2- I tried to create an autocomplete.yaml but I don't know yet the right syntax. I will test custom items when we have the documentation :)

> Show SC version number in settings, and a link to the changelog (#112).

Working perfect! 

> Settings: Variables are now located on their own tab (#110).

My only suggestion is to always let the `Variables` tab as the last tab in the UI. 

My felling as a user is that we will always use the variables tab less than the other tabs 


Nothing important, just a suggestion :)


---

Have a great day!!!!!!!!!

**Reply by Taitava** [reply]
> I love the attention that you have with {{tags:separator}} to insert the caret position inside the }} instead of outside for variables with separator 😍

Thank you! 🙂 As I use an autocomplete menu a lot in programming anyway, I have the habit to think about where a user will probably type text after autocompleting something. And, if you ever notice the caret going to be in a less optimal place after autocompletion, please let me know and I'll think about tweaking it more. 👍 

> 1- Show Autocomplete menu -> link `info` button
> 
> The link from this button https://publish.obsidian.md/shellcommands/Variables/Autocomplete/Autocomplete is 404 File Not Found

Yea, I actually thought I would publish that page later when `0.8.0` gets released (so that users won't get confused when the documentation shows a feature that does not exist in the latest release). But it's more important to get the instructions for you, so I'll publish that page now.

If you'll still see a _404 not found_ error, it's a cache problem and will likely go away after some time, but you can also use another browser, put your browser in privacy mode, or use another machine to access that link without old cache causing problems. This affects all Obsidian Publish sites, all changes to them may take some time to appear due to caching.

> My only suggestion is to always let the `Variables` tab as the last tab in the UI.
> 
> My felling as a user is that we will always use the variables tab less than the other tabs

Thanks for the suggestion, I will change it! If nothing important comes up, you'll see it in the final release version of `0.8.0`, but if something else comes up, you'll see it in a third beta, but so far it might be that a third beta is not needed. 👍

**Reply by Taitava** [reply]
I forgot to publish autocomplete related images, but now they are published, too. This also has the cache issue: if images don't show up, put your browser in privacy mode (or close a private window and reopen a private window, if you had one open already).

**Reply by FelipeRearden** [reply]
@Taitava 

~~Before I answer I just wanna confirm something strange that I see ...~~

~~Could you create for me an autocomplete.yaml right with these two commands~~

Thanks @Taitava !

The examples in the Documentation solve my doubt !!!!

**Reply by FelipeRearden** [reply]
> if you ever notice the caret going to be in a less optimal place after autocompletion, please let me know and I'll think about tweaking it more. 👍 

I will @Taitava !!! Thanks :)

> Yea, I actually thought I would publish that page later when `0.8.0` gets released (so that users won't get confused when the documentation shows a feature that does not exist in the latest release). But it's more important to get the instructions for you, so I'll publish that page now.
>
> If you'll still see a _404 not found_ error, it's a cache problem and will likely go away after some time, but you can also use another browser, put your browser in privacy mode, or use another machine to access that link without old cache causing problems. This affects all Obsidian Publish sites, all changes to them may take some time to appear due to caching.

Thanks @Taitava !!!! It's perfect now :)

> Thanks for the suggestion, I will change it! If nothing important comes up, you'll see it in the final release version of `0.8.0`, but if something else comes up, you'll see it in a third beta, but so far it might be that a third beta is not needed. 👍 

Great ! @Taitava thanks !!!!!!

**Reply by FelipeRearden** [reply]
@Taitava 

Just one little thing that I saw. I think you don't have to do anything it is just for our future reference

SC behavior with autocomplete.yaml file in the SC folder

1- autocomplete.yaml in empty:

notification shows that file needs to be filled -> unable to parse

2- autocomplete.yaml is not empty + the syntax is wrong:

The SC setting tab disappear from the Settings windows

Before below RSS Reader

![1](https://user-images.githubusercontent.com/85180696/145576061-05f876ba-283c-45db-a2cf-e24a7e5ba1b4.png)

After (below RSS Reader)

![2](https://user-images.githubusercontent.com/85180696/145576471-eb1b1ca0-37cd-46ac-859f-afcad6f52ede.png)


I know that SC is installed because I CAN see the commands in the Command Palette 

---

Using the right syntax, everything is perfect @Taitava !!!!

**Reply by Taitava** [reply]
> 1- autocomplete.yaml in empty:
> 
> notification shows that file needs to be filled -> unable to parse

I added now a new check for an empty file content, so now the error message is even more clear telling that loading failed because content is empty.

> 2- autocomplete.yaml is not empty + the syntax is wrong:
> 
> The SC setting tab disappear from the Settings windows

Thanks for reporting! I fixed this:
1. Now there is a visible error message, which tries to give some kind of hint about what's wrong in the file. This hint comes from Obsidian's API (`parseYaml()` function), and I do not know all the possible syntax error messages that come from there, but at least one I got from there looks helpful.
2. Now the whole plugin does not crash to this kind of error situation, so the plugin can continue to work, and even the normal variable autocomplete should work, but without any content from `autocomplete.yaml`.

I'll create a new beta version.

**Comment by Taitava** [comment]
# 0.8.0-beta.3 is available

Quite small changes:
- Fixed a crash if autocomplete.yaml could not be parsed.
- Settings: Variables tab is now moved to be the last one.

**Reply by Taitava** [reply]
@FelipeRearden can you try again some syntax errors with autocomplete.yaml and check that the SC settings view will not disappear, and that error messages look understandable.

If the error handling works well, I think I can then release `0.8.0`. Thank you for helping me to improve this, I appreciate it! 😎

**Reply by FelipeRearden** [reply]
> - Fixed a crash if autocomplete.yaml could not be parsed.

100% working perfect @Taitava !!!!!!

> - Settings: Variables tab is now moved to be the last one.

100% working perfect ! Thanks for the change @Taitava !

> @FelipeRearden can you try again some syntax errors with autocomplete.yaml and check that the SC settings view will not disappear, and that error messages look understandable.

100% working perfect @Taitava !

- autocomplete empty
- autocomplete right
- autocomplete wrong 

The notifications messages are crystal clear for me :)

> If the error handling works well, I think I can then release `0.8.0`. Thank you for helping me to improve this, I appreciate it! 😎 


I think we are ready for the merge @Taitava !!!!!!

Have a great day!!!!!

**Reply by Taitava** [reply]
Thank you! I'll close this beta test now and release `0.8.0`. Actually, I just relialized I didn't even remember to do a change to manifest-beta.json that would have allowed the third beta to be installed via the BRAT plugin. But that does not matter as you tend to use manual installation for betas anyway. 🙂 Btw is there any problem caused by a missing manifest.json file when you do manual beta installations? If it causes problems, I think I can make a manifest.json file for future betas, but I just have to make sure I don't accidentally commit it to the main branch (= that would accidentally release a beta as a normal release).

**Reply by FelipeRearden** [reply]
> Btw is there any problem caused by a missing manifest.json file when you do manual beta installations? If it causes problems, I think I can make a manifest.json file for future betas, but I just have to make sure I don't accidentally commit it to the main branch (= that would accidentally release a beta as a normal release).

@Taitava didn't felt any problem at my end. The only thing that the version number isn't update (which is caused by the absence of the manifest.json)

Feel free to continue this way. I think that BRAT is the right for every user/tester. 

I don't installed yet in my test vault because I didn't like the way that the BRAT developer treats me when I was testing a another beta plugin from him. I spend a lot of time testing and suggesting FRs and I didnt get any feedback/attention from him: 
- maybe is because I'm not a VIP in the community; 
- maybe is because I don't speak English;
- maybe he thinks that I'm dumb.    

This situation made me very very very sad.

This way I don't wanna use BRAT to get risk of finding an issue, report and get sad again. 

But I think BRAT is the right way for testers :)

**Reply by Taitava** [reply]
Ah, sad that you've had bad experiences 🙁 . Sure, you can do the testing your way, no need to use BRAT.

**Reply by FelipeRearden** [reply]
Thanks @Taitava and sorry for the vent.

All good ! Lets keep making this amazing plug-in even more powerful :)

**Reply by Taitava** [reply]
No problem, you can always tell about negative things, too. 👍

**Reply by FelipeRearden** [reply]
Thanks @Taitava !!!!! I really appreciate :) :) :)
